lib/beautiful/gtk.lua

```diff
@@ -0,0 +1,169 @@
+
+
+local gtk = {
+  cached_theme_variables = nil
+}
+
+
+local function convert_gtk_channel_to_hex(channel_value)
+    return string.format("%02x", gears_math.round(channel_value * 255))
+end
+
+local function convert_gtk_color_to_hex(gtk_color)
+  return "#" ..
+    convert_gtk_channel_to_hex(gtk_color.red) ..
+    convert_gtk_channel_to_hex(gtk_color.green) ..
+    convert_gtk_channel_to_hex(gtk_color.blue) --..
+  --convert_gtk_channel_to_hex(gtk_color.alpha)
+end
+
+local function lookup_gtk_color_to_hex(_style_context, color_name)
+  local gtk_color = _style_context:lookup_color(color_name)
+  if not gtk_color then
+    return nil
+  end
+  return convert_gtk_color_to_hex(gtk_color)
+end
+
+local function get_gtk_color_property_to_hex(_style_context, property_name)
+    local state = _style_context:get_state()
+    local property = _style_context:get_property(property_name, state)
+    if not property then
+      return nil
+    end
+    local result = convert_gtk_color_to_hex(property.value)
+    property:unset()
+    return result
+end
+
+local function read_gtk_color_properties_from_widget(gtk_widget, properties)
+  local _style_context = gtk_widget:get_style_context()
+  local result = {}
+  for result_key, style_context_property in pairs(properties) do
+    result[result_key] = get_gtk_color_property_to_hex(
+        _style_context, style_context_property
+    )
+  end
+  return result
+end
+
+
+--- Get GTK+3 theme variables from GtkStyleContext
+-- @treturn table subj
+function gtk.get_theme_variables()
+  if gtk.cached_theme_variables then
+    return gtk.cached_theme_variables
+  end
+
+  local result = {}
+  local lgi = require('lgi')
+  local Gtk = lgi.Gtk
+  local window
+  pcall(function()
+    window = Gtk.Window{
+      --on_destroy = Gtk.main_quit,
+    }
+  end)
+  if not window then
+    gears_debug.print_warning("Seems like GTK+3 theme is not set correctly or luajit was built with an incomptible GTK+3 version.")
+    return nil
+  end
+  --window:set_override_redirect(true)
+  local style_context = window:get_style_context()
+
+  local font = style_context:get_font(Gtk.StateFlags.NORMAL)
+  --result.font_family = font:get_family()
+  --result.font_size = font:get_size()/1024
+  result.font = font:to_string()
+
+  result = join(result, read_gtk_color_properties_from_widget(
+    window, {
+      bg_color="background-color",
+      fg_color="color",
+    }
+  ))
+  result = join(result, read_gtk_color_properties_from_widget(
+    Gtk.Entry(), {
+      base_color="background-color",
+      text_color="color",
+    }
+  ))
+  result = join(result, read_gtk_color_properties_from_widget(
+    Gtk.ToggleButton(), {
+      selected_bg_color="background-color",
+      selected_fg_color="color",
+    }
+  ))
+  local button = Gtk.Button()
+  result = join(result, read_gtk_color_properties_from_widget(
+    button, {
+      button_bg_color="background-color",
+      button_fg_color="color",
+    }
+  ))
```

Comment:
Thanks, I find it a lot easier to figure out what the above code does than in the old version